src/toast/cc-toaster.js

```diff
@@ -0,0 +1,432 @@
+      _toasts: { type: Array, attribute: false },
+    };
+  }
+
+  constructor () {
+    super();
+
+    /** @type {ToastAnimation} Animation to be played when a toast appears.*/
+    this.animation = 'fade';
+
+    /** @type {number|null} The maximum number of toasts that can be displayed at the same time. null for no limit.*/
+    this.maxToasts = null;
+
+    /** @type {ToastPosition} Indicates where the toaster is placed. This property controls only how the toasts will be displayed (not where).*/
+    this.position = 'top-right';
+
+    /** @type {ToastConfig|null} The default config to apply to toasts. Changing this property takes effect on new toasts only. The properties of a default configuration are always used as fallback. */
+    this.toastDefaultConfig = null;
+
+    /** @type {Array<Toast>} Array of toasts currently displayed. This property should not be used directly. Instead, you should dispatch a `cc-toast:show` event. */
+    this._toasts = [];
+
+    this._dismissingSet = new Set();
+  }
+
+  /* region PRIVATE METHODS */
+  /**
+   * Based on the given toast, forges a new toast where the default config is applied, and where the key is generated if not present.
+   *
+   * @param {Toast} toast
+   * @returns {Toast}
+   * @private
+   */
+  _completeToast (toast) {
+    return {
+      ...toast,
+      key: toast.key ?? Math.random().toString(36).slice(2),
+      config: {
+        ...this.toastDefaultConfig,
+        ...toast.config,
+      },
+    };
+  }
+
+  /**
+   * @param {string} key
+   * @returns {Toast}
+   * @private
+   */
+  _getToastByKey (key) {
+    return this._toasts.find((t) => t.key === key);
+  }
+
+  /**
+   * Shows a new toast.
+   *
+   * If the given toast doesn't contain a `key` property, a random key is generated automatically.
+   * If a toast with the given key already exists, nothing happens.
+   * If the maxToasts is reached, the oldest toast will be dismissed automatically.
+   * If the maxToasts is 0, nothing happens.
+   *
+   * @param {Toast} toast the toast to show
+   * @returns {Toast|undefined} the completed toast. Returns undefined if no toast was shown.
+   */
+  _show (toast) {
+    if (this.maxToasts === 0) {
+      return;
+    }
+
+    // dismiss last toast if max is reached
+    if (this.maxToasts) {
+      const activeToasts = this._toasts.filter((toast) => this._isDismissing(toast.key) !== true);
+
+      if (activeToasts.length >= this.maxToasts) {
+        this._dismiss(activeToasts[activeToasts.length - 1].key);
+      }
+    }
+
+    const completedToast = this._completeToast(toast);
+
+    // do nothing if a toast with the same key already exists.
+    if (this._toasts.find((toast) => toast.key === completedToast.key)) {
+      return;
+    }
+
+    this._toasts = [completedToast, ...this._toasts];
+
+    return completedToast;
+  }
+
+  _dismiss (key) {
+    dispatchCustomEvent(this, 'cc-toast:dismiss', { key, timeoutExpired: false });
+  }
+
+  _startHide (key) {
```

Comment:
I'm not sure but it seems like we use `hide` and `dismiss` for the same thing. Maybe we don't...